volatility3/framework/plugins/windows/skeleton_key_check.py

```diff
@@ -0,0 +1,578 @@
+                proc_id = proc.UniqueProcessId
+                proc_layer_name = proc.add_process_layer()
+            except exceptions.InvalidAddressException as excp:
+                vollog.debug("Process {}: invalid address {} in layer {}".format(proc_id, excp.invalid_address,
+                                                                                 excp.layer_name))
+                continue
+
+            proc_layer = self.context.layers[proc_layer_name]
+
+            for vad in proc.get_vad_root().traverse():
+                filename = vad.get_file_name()
+                if type(filename) == renderers.NotApplicableValue or not filename.lower().endswith("cryptdll.dll"):
+                    continue
+       
+                cryptdll_base = vad.get_start()
+                cryptdll_size = vad.get_end() - cryptdll_base
+
+                break
+
+            lsass_proc = proc
+            break
+
+        return lsass_proc, proc_layer_name, cryptdll_base, cryptdll_size
+
+    def _find_csystems_with_symbols(self, proc_layer_name: str, 
+                                          cryptdll_types: interfaces.context.ModuleInterface, 
+                                          cryptdll_base: int, 
+                                          cryptdll_size: int) -> \
+                                          Tuple[interfaces.objects.ObjectInterface, int, int]:
+        """
+        Attempts to find CSystems and the expected address of the handlers.
+        Relies on downloading and parsing of the cryptdll PDB file.
+
+        Args:
+            proc_layer_name: the name of the lsass.exe process layer
+            cryptdll_types: The types from cryptdll binary analysis
+            cryptdll_base: the base address of cryptdll.dll
+            crytpdll_size: the size of the VAD for cryptdll.dll
+        
+        Returns:
+            A tuple of:
+            array: An initialized Volatility array of _KERB_ECRYPT structures
+            rc4HmacInitialize: The expected address of csystem Initialization function 
+            rc4HmacDecrypt: The expected address of the csystem Decryption function
+        """
+        try:
+            cryptdll_symbols = pdbutil.PDBUtility.symbol_table_from_pdb(self.context, 
+                                                                    interfaces.configuration.path_join(self.config_path, 'cryptdll'),
+                                                                    proc_layer_name,
+                                                                    "cryptdll.pdb",
+                                                                    cryptdll_base,
+                                                                    cryptdll_size)
+        except exceptions.VolatilityException:
+            return None, None, None
+
+        array_start, count, rc4HmacInitialize, rc4HmacDecrypt = \
+            self._find_array_with_pdb_symbols(cryptdll_symbols, cryptdll_types, proc_layer_name, cryptdll_base) 
```

Comment:
I am condensing the array building code across functions, but I am not sure how I would use object_from_symbol for the array. We have _cryptdll_symbols_ which is from the PDB and only has global variable addresses (no type info), and then we have _cryptdll_types_ which is the hand-written JSON from my reversing. I am not sure how to combine those two in a way to then be able to replace manually constructing the array of _KERB_ECRYPT_ instances. If there is a way to do this then I would definitely like to learn how.

Reply:
Ok, I see.  Yeah, if the symbols are from the PDB and they don't include a type, then it's not going to end up saving much time to build it any other way.  Still seems a bit odd to be returning the `array_start` and using that here, rather than constructing the array in the `_find_array_with_pdb_symbols` call.  It feels like the function call was designed just for this one use, I'd have thought it would be more useful to just create the array in the call and return it here?  I can't see an occasion where you'd want just the offset, rather than the whole array...

I figured out what's going on, and I think it might cause confusion ugliness down the road.  5:S  I think you were expecting to be using the `symbol_shift` which you can apply when loading the symbol table, which means that all symbols loaded come out with that offset (that's how linux handles ASLR).  I think we might've built up the ability for confusion, because if you've shifted the symbols and provide the right offset for the module, when you construct an object that way, it'll add the module offset on *again*.  So you've got to use one or the other, but not both...  5:S  I knew we were headed into dark waters when I added that feature...  Needs some thinking about.  For now, any solution you decide to go for (three additions, or building the symbol table with a `symbol_shift`) will be fine, you just need to be careful about building a module and handing in an offset to it...